File: genetic_algorithm/pid_tuning.py
from deap import creator, base, tools, algorithms
import random
import struct
from control.matlab import *
import numpy as np
import pdb
import sys
np.seterr(over='ignore')

def bin_to_float32(binary_string):
    f = int(binary_string, 2)
    float32 = struct.unpack('f', struct.pack('I', f))[0]
    return float32

creator.create("fitness", base.Fitness, weights=(-1.0,))
creator.create("individual",list,fitness=creator.fitness)

toolbox = base.Toolbox()
toolbox.register('attr_bool',random.randint,0,1)
toolbox.register('individual',tools.initRepeat,creator.individual, toolbox.attr_bool,n=32*3)
toolbox.register('population',tools.initRepeat,list,toolbox.individual)

Q = 1 
R = 0.01
dt = 0.1
t = np.arange(0,20,dt)
s = tf('s')
G = 3/(s**2+s+1)
FLT_MAX = sys.float_info.max
def evalLQ(individual):
    Kp_binary = individual[:32]
    Ki_binary = individual[32:64]
    Kd_binary = individual[64:]
    Kp_binary_string = ''.join([str(k) for k in Kp_binary])
    Ki_binary_string = ''.join([str(k) for k in Ki_binary])
    Kd_binary_string = ''.join([str(k) for k in Kd_binary])
    Kp_float32 = bin_to_float32(Kp_binary_string) 
    Ki_float32 = bin_to_float32(Ki_binary_string) 
    Kd_float32 = bin_to_float32(Kd_binary_string) 
    K = Kp_float32 + Ki_float32/s + Kd_float32*s/(1+0.1*s)
    loop = series(K,G)
    closedLoop = feedback(loop, 1)
    try:
        y,_ = step(closedLoop,t)
    except Exception:
        J = FLT_MAX
        return J,
    u,_,_ = lsim(K, 1-y, t)
    J = dt*sum(Q*(1-y)**2 + R*u**2)
    if np.isnan(J):
        J = FLT_MAX
    return J,

toolbox.register('evaluate',evalLQ)
toolbox.register('mate',tools.cxTwoPoint)
toolbox.register('mutate',tools.mutFlipBit,indpb=0.05)
toolbox.register('select',tools.selTournament,tournsize=20)

def printResult(individual):
    bestFit = individual.fitness.values[0]
    Kp_binary = individual[:32]
    Ki_binary = individual[32:64]
    Kd_binary = individual[64:]
    Kp_binary_string = ''.join([str(k) for k in Kp_binary])
    Ki_binary_string = ''.join([str(k) for k in Ki_binary])
    Kd_binary_string = ''.join([str(k) for k in Kd_binary])
    Kp_float32 = bin_to_float32(Kp_binary_string) 
    Ki_float32 = bin_to_float32(Ki_binary_string) 
    Kd_float32 = bin_to_float32(Kd_binary_string) 
    print("Kp:{}, Ki:{}, Kd:{}, J:{}".format(Kp_float32,Ki_float32,Kd_float32,bestFit))

population = toolbox.population(n=100)
GEN = 300
CXPB = 0.8
MUTPB = 0.1
for gen in range(GEN):
    offspring = algorithms.varAnd(population, toolbox, cxpb=CXPB, mutpb=MUTPB)
    fits = toolbox.map(toolbox.evaluate, offspring)
    for fit, ind in zip(fits, offspring):
        ind.fitness.values = fit
    population = toolbox.select(offspring, k=len(population))
    best = tools.selBest(population, 1)[0]
    bestFit = best.fitness.values[0]
    # print(gen, bestFit)
    print(gen,end=' ')
    printResult(best)

File: genetic_algorithm/pid_tuning.m
clear; clc; close all;
s=tf('s');
G=3/(s^2+s+1);
K=pidtune(G,'pid')
loop=series(K,G);
closedLoop=feedback(loop,1);

% GEN100 Kp:1.3103935718536377, 
% Ki:0.9094977378845215, Kd:0.6930094361305237, 
%J:0.39028349787582434
K_GA = 1.31+0.909/s+0.693*s/(1+0.1*s);
[y_ga,t2] = step(feedback(K_GA*G,1));

sys = ss(G);
A=sys.A;
B=sys.B;
C=sys.C;
D=sys.D;

Q = 1;
R = 0.1;
N = 0;
[K,S,e] = lqi(sys,Q,R,N);
K1 = K(1:2);
K2 = K(3);
Ae = [A-B*K1 -B*K2
     -C 0];
Be = [B;1];
Ce = [C 0];
De = D;
sys2 = ss(Ae,Be,Ce,De);
[y3,t]=step(sys2,t2);

[y,t]=step(closedLoop,t2);
[y0,t]=step(G,t2);

plot(t,y,'b',t2,y_ga,'r',t2,y3,'k');
legend('MATLAB PID Tuner','GA','LQI')

big;

